provider/kubernetes/weight_allocator.go

```diff
@@ -0,0 +1,163 @@
+	servicePercentageWeights, err := getServicesPercentageWeights(ingress)
+	if err != nil {
+		return nil, err
+	}
+
+	allocator := &fractionalWeightAllocator{
+		serviceWeights: map[ingressService]int{},
+	}
+
+	serviceInstanceCounts, err := getServiceInstanceCounts(ingress, client)
+	if err != nil {
+		return nil, err
+	}
+
+	serviceWeights := map[ingressService]int{}
+	for _, rule := range ingress.Spec.Rules {
+		// key: rule path string
+		// value: service names
+		fractionalPathServices := map[string][]string{}
+
+		// key: rule path string
+		// value: fractional percentage weight
+		fractionalPathWeights := map[string]percentageValue{}
+
+		for _, pa := range rule.HTTP.Paths {
+			if _, ok := fractionalPathWeights[pa.Path]; !ok {
+				fractionalPathWeights[pa.Path] = newPercentageValueFromFloat64(1)
+			}
+
+			if weight, ok := servicePercentageWeights[pa.Backend.ServiceName]; ok {
+				ingSvc := ingressService{
+					host:    rule.Host,
+					path:    pa.Path,
+					service: pa.Backend.ServiceName,
+				}
+
+				serviceWeights[ingSvc] = weight.computeWeight(serviceInstanceCounts[ingSvc])
+
+				fractionalPathWeights[pa.Path] = fractionalPathWeights[pa.Path].sub(weight)
+				if fractionalPathWeights[pa.Path].toFloat64() < 0 {
+					return nil, fmt.Errorf("percentage value %s out of range", fractionalPathWeights[pa.Path].String())
+				}
+			} else {
+				fractionalPathServices[pa.Path] = append(fractionalPathServices[pa.Path], pa.Backend.ServiceName)
+			}
+		}
+
+		for pa, svcs := range fractionalPathServices {
+			totalFractionalInstanceCount := 0
+
+			for _, svc := range svcs {
+				totalFractionalInstanceCount += serviceInstanceCounts[ingressService{
+					host:    rule.Host,
+					path:    pa,
+					service: svc,
+				}]
+			}
+
+			for _, svc := range svcs {
+				ingSvc := ingressService{
+					host:    rule.Host,
+					path:    pa,
+					service: svc,
+				}
+				serviceWeights[ingSvc] = fractionalPathWeights[pa].computeWeight(totalFractionalInstanceCount)
+			}
+		}
+	}
+
+	allocator.serviceWeights = serviceWeights
+	return allocator, nil
+}
+
+func (f *fractionalWeightAllocator) getWeight(host, path, serviceName string) int {
+	return f.serviceWeights[ingressService{
+		host:    host,
+		path:    path,
+		service: serviceName,
+	}]
+}
+
+func getServiceInstanceCounts(ingress *extensionsv1beta1.Ingress, client Client) (map[ingressService]int, error) {
+	serviceInstanceCounts := map[ingressService]int{}
+
+	for _, rule := range ingress.Spec.Rules {
+		for _, pa := range rule.HTTP.Paths {
+			count := 0
+			endpoints, exists, err := client.GetEndpoints(ingress.Namespace, pa.Backend.ServiceName)
+			if err != nil {
+				return nil, fmt.Errorf("fail to get endpoints %s/%s: %v", ingress.Namespace, pa.Backend.ServiceName, err)
+			}
+			if !exists {
+				return nil, fmt.Errorf("fail to get endpoints %s/%s: non-existent endpoint", ingress.Namespace, pa.Backend.ServiceName)
```

Comment:
Can we rephrase the text to `Endpoints not found for %s/%s` to be consistent with our [existing message](https://github.com/containous/traefik/blob/bddb4cc33c87108d6bc9ce1dc456024031dfe63b/provider/kubernetes/kubernetes.go#L308)?